File: easy/ValidPalindrome.cpp
#include <algorithm>
#include <cctype>
#include <iostream>

using namespace std;

class Solution {
public:
  bool isPalindrome(string s) {
    string filter;
    for (auto ch : s) {
      if (isalnum(ch)) {
        filter += tolower(ch);
      }
    }

    string reversed = filter;
    reverse(reversed.begin(), reversed.end());
    return reversed == filter;
  }
};

int main() {

  string str = "fr AM Arf";
  Solution *s = new Solution();
  cout << s->isPalindrome(str) << "yes";
}

File: ts/easy/ValidPalindrome.ts
class Solution {
	isPalindrome(s: string): boolean {
		s = s.toLowerCase();
		let [l, r] = [0, s.length - 1];
		let specialChar = /\W|_/;

		while (l < r) {
			if (specialChar.test(s[l])) {
				l++;
			} else if (specialChar.test(s[r])) {
				r--;
			} else if (s[l] !== s[r]) {
				return false
			} else {
				r--;
				l++;
			}
		}

		return true;
	}
}
